File: solutions/elixir/anagram/3/lib/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    base = String.downcase(base)
    Enum.filter(candidates, fn candidate ->
      candidate = String.downcase(candidate)
      candidate !== base && count_letters(candidate) === count_letters(base)
    end)
  end

  defp count_letters(word) do
    word
    |> String.graphemes()
    |> Enum.reduce(%{}, fn letter, acc -> Map.update(acc, letter, 1, &(&1 + 1)) end)
  end
end

File: solutions/ocaml/anagram/3/anagram.ml
open Base

let count_letters word =
  word
  |> String.to_list
  |> List.fold
       ~init:(Map.empty (module Char))
       ~f:(fun map letter ->
         Map.update map letter ~f:(function
           | None -> 1
           | Some n -> n + 1))

let anagrams base candidates =
  let base = String.lowercase base in
  List.filter candidates ~f:(fun c ->
      let c = String.lowercase c in
      (not (String.equal c base))
      && Map.equal Int.equal (count_letters c) (count_letters base))
